oximeter/impl/src/schema/ir.rs

```diff
@@ -0,0 +1,1288 @@
+enum CurrentVersion {
+    Active { version: NonZeroU8, fields: BTreeSet<String> },
+    Inactive { removed_in: NonZeroU8 },
+}
+
+impl CurrentVersion {
+    fn version(&self) -> NonZeroU8 {
+        match self {
+            CurrentVersion::Active { version, .. } => *version,
+            CurrentVersion::Inactive { removed_in } => *removed_in,
+        }
+    }
+
+    fn fields(&self) -> Option<&BTreeSet<String>> {
+        match self {
+            CurrentVersion::Active { fields, .. } => Some(fields),
+            CurrentVersion::Inactive { .. } => None,
+        }
+    }
+}
+
+fn defn_error<T>(s: String) -> Result<T, MetricsError> {
+    Err(MetricsError::SchemaDefinition(s))
+}
+
+/// Load the list of timeseries schema from a schema definition in TOML format.
+pub fn load_schema(
+    contents: &str,
+) -> Result<Vec<TimeseriesSchema>, MetricsError> {
+    toml::from_str::<TimeseriesDefinition>(contents)
+        .map_err(|e| MetricsError::Toml(e.to_string()))
+        .and_then(TimeseriesDefinition::into_schema_list)
+}
+
+// Find schema of a specified version in an iterator, or the latest.
+pub(super) fn find_schema_version(
+    list: impl Iterator<Item = TimeseriesSchema>,
+    version: Option<NonZeroU8>,
+) -> Vec<TimeseriesSchema> {
+    match version {
+        Some(ver) => list.into_iter().filter(|s| s.version == ver).collect(),
+        None => {
+            let mut last_version = BTreeMap::new();
+            for schema in list {
+                let metric_name = schema.component_names().1.to_string();
+                match last_version.entry(metric_name) {
+                    Entry::Vacant(entry) => {
+                        entry.insert((schema.version, schema.clone()));
+                    }
+                    Entry::Occupied(mut entry) => {
+                        let existing_version = entry.get().0;
+                        if existing_version < schema.version {
+                            entry.insert((schema.version, schema.clone()));
+                        }
+                    }
+                }
+            }
+            last_version.into_values().map(|(_ver, schema)| schema).collect()
+        }
+    }
+}
+
+fn extract_authz_scope(
+    metric_name: &str,
+    authz_scope: AuthzScope,
+    field_schema: &BTreeSet<FieldSchema>,
+) -> Result<AuthzScope, MetricsError> {
+    let check_for_key = |scope: &str| {
+        let key = format!("{scope}_id");
+        if field_schema.iter().any(|field| {
+            field.name == key && field.field_type == FieldType::Uuid
+        }) {
+            Ok(())
+        } else {
+            defn_error(format!(
+                "Metric '{}' has '{}' authorization scope, and so must \
+                contain a field '{}' of UUID type",
+                metric_name, scope, key,
+            ))
+        }
+    };
+    match authz_scope {
+        AuthzScope::Silo => check_for_key("silo")?,
+        AuthzScope::Project => check_for_key("project")?,
+        _ => {}
```

Comment:
Nit - could we replace `_` with an explicit list of variants, so new variants added in the future have to be considered?

Reply:
Yeah, good call, will update.

Reply:
Done in dca0355f3